src/Microsoft.Azure.WebJobs.Host/Bindings/BindingFactory.cs

```diff
@@ -136,60 +136,60 @@ public IBindingProvider BindToGenericValueProvider<TAttribute>(Func<TAttribute,
+        /// <param name="builderType">A class with builder methods on it. This will get instantiated to perform the builder rule.</param>
+        /// <param name="constructorArgs">constructor arguments to pass to the typeBuilder instantiation. This can be used 
+        /// to flow state (like configuration, secrets, etc) from the configuration to the specific binding</param>
+        /// <returns>A binding rule.</returns>
+        public IBindingProvider BindToInput<TAttribute, TType>(
+            bool allowConversions,
+            Type builderType,
+            params object[] constructorArgs)
+                where TAttribute : Attribute
         {
-            return this.BindToExactAsyncType<TAttribute, TUserType>((attr) => Task.FromResult(buildFromAttribute(attr)));
+            var cm = allowConversions ? this._converterManager : null;
+            var pm = PatternMatcher.New(builderType, constructorArgs);
+            return new BindToInputBindingProvider<TAttribute, TType>(this._nameResolver, cm, pm);
         }
 
         /// <summary>
-        /// Create a binding provider that binds the parameter to an specific instance of TUserType. 
+        /// General rule for binding to an input type for a given attribute. 
         /// </summary>
-        /// <typeparam name="TAttribute">type of binding attribute on the user's parameter.</typeparam>
-        /// <typeparam name="TUserType">The exact type of the user's parameter that this will bind to.</typeparam>
-        /// <param name="buildFromAttribute">builder function to create the object that will get passed to the user function.</param>
-        /// <param name="buildParameterDescriptor">An optional function to create a specific ParameterDescriptor object for the dashboard. If missing, a default ParameterDescriptor is created. </param>
-        /// <param name="postResolveHook">an advanced hook for translating the attribute. </param>
-        /// <returns>A binding provider that applies these semantics.</returns>
-        public IBindingProvider BindToExactAsyncType<TAttribute, TUserType>(
-            Func<TAttribute, Task<TUserType>> buildFromAttribute,
-            Func<TAttribute, ParameterInfo, INameResolver, ParameterDescriptor> buildParameterDescriptor = null,
-            Func<TAttribute, ParameterInfo, INameResolver, Task<TAttribute>> postResolveHook = null)
+        /// <typeparam name="TAttribute">Type of binding attribute on the user's parameter.</typeparam>
+        /// <typeparam name="TType">The user type must be compatible with this type for the binding to apply.</typeparam>
+        /// <param name="allowConversions">
+        /// If false, then the user parameter type must exactly match TType. 
+        /// If true, this rule can use the ConverterManager to convert from TType to the final parameter type </param>
+        /// <param name="builderInstance">Instance with converter methods on it.</param>
```

Comment:
It's not an instance with "converter" methods, is it? Shouldn't it be an instance with "builder" methods on it? Also ties in with my comment above.